### PR TITLE
Fixed #1101 : Implement Query API

### LINKED ISSUE #1101: 2.0: Query API
Spec: https://github.com/couchbaselabs/couchbase-lite-apiv2/tree/master/query/java

### PR DESCRIPTION
**Pull request by pasin**
#1101

### CODE REVIEW

File: shared/src/main/java/com/couchbase/lite/Expression.java [hunk 1]
```diff
@@ -0,0 +1,586 @@
...
+    public static Expression negated(Object expression) {
+        return new CompoundExpression(Arrays.asList(expression), CompoundExpression.OpType.Not);
+    }
+
+    /**
+     * Create a negated expression to represent the negated result of the given expression.
+     * @param expression the expression to be negated.
+     * @return a negated expression.
+     */
+    public static Expression not(Object expression) {
```
Comment (hideki) [inline shared/src/main/java/com/couchbase/lite/Expression.java]: I feel one of negate(...) or not(...) methods could be enough.

File: shared/src/main/java/com/couchbase/lite/Expression.java [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+    }
+
+    /**
+     * Create a between expression that evaluates whether or not the current expression is
+     * between the given expressions.
+     * @param expression1 the lower bound expression.
+     * @param expression2 the upper bound expression.
+     * @return a between expression.
+     */
+    public Expression between(Object expression1, Object expression2) {
```
Comment (hideki) [inline shared/src/main/java/com/couchbase/lite/Expression.java]: inclusive??

File: shared/src/main/java/com/couchbase/lite/Expression.java [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+    }
+
+    static class BinaryExpression extends Expression {
+        enum OpType {
+            Add, Between, Divide, EqualTo, GreaterThan, GreaterThanOrEqualTo,
+            In, Is, IsNot, LessThan, LessThanOrEqualTo, Like, Matches,
+            Modulus, Multiply, NotEqualTo, Subtract, RegexLike,
+        }
+
+        private Object lhs;
```
Comment (hideki) [inline shared/src/main/java/com/couchbase/lite/Expression.java]: lhs and rhs: It is hard to guess what this variables indicate. I can guess left and right.

File: shared/src/main/java/com/couchbase/lite/ResultSet.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+package com.couchbase.lite;
+
+import com.couchbase.litecore.C4QueryEnumerator;
+import com.couchbase.litecore.LiteCoreException;
+
+public class ResultSet {
```
Comment (hideki) [inline shared/src/main/java/com/couchbase/lite/ResultSet.java]: Is `ResultSet` not `Iterable`?

File: android/CouchbaseLite/src/androidTest/java/com/couchbase/lite/QueryTest.java [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+        return numbers;
+    }
+
+    private void runTestWithNumbers(List<Map<String, Object>> numbers, Object[][] cases)
+            throws Exception {
+        for (Object[] c : cases) {
+            Expression w = (Expression) c[0];
+            String[] docIDs = (String []) c[1];
+            final List<String> docIDList = new ArrayList<String>(Arrays.asList(docIDs));
+            Query q = Query.select().from(DataSource.database(db)).where(w);
```
Comment (pasin) [inline android/CouchbaseLite/src/androidTest/java/com/couchbase/lite/QueryTest.java]: That is from the spec. In the future, we could have more than database data source.

File: android/CouchbaseLite/src/androidTest/java/com/couchbase/lite/QueryTest.java [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+                    if (docIDList.contains(docID))
+                        docIDList.remove(docID);
+                }
+            });
+            assertEquals(0, docIDList.size());
+            assertEquals(docIDs.length, rows);
+        }
+    }
+
+    private String[] $docids(int...numbers) {
```
Comment (pasin) [inline android/CouchbaseLite/src/androidTest/java/com/couchbase/lite/QueryTest.java]: It's just a language kimick for generating somethings. I borrowed the $ from objc.

File: shared/src/main/java/com/couchbase/lite/DataSource.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    /* package */ Object getSource() {
+        return this.source;
+    }
+
+    /**
+     * Database as a data source for query.
+     */
+    public static class Database extends DataSource {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/DataSource.java]: This is also from spec.

File: shared/src/main/java/com/couchbase/lite/Expression.java (hunk 1 above)
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Expression.java]: This is also from the spec.

File: shared/src/main/java/com/couchbase/lite/Expression.java (hunk 3 above)
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Expression.java]: left hand side and right hand side.

File: shared/src/main/java/com/couchbase/lite/ResultSet.java (hunk 4 above)
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/ResultSet.java]: The spec is using ResultSet. Java cannot also implement Iterator as it cannot implement hasNext().

File: shared/src/main/java/com/couchbase/lite/Expression.java (hunk 2 above)
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Expression.java]: Yes.

File: shared/src/main/java/com/couchbase/lite/Expression.java (hunk 3 above)
Comment (hideki) [inline shared/src/main/java/com/couchbase/lite/Expression.java]: I prefer clean/clear variable naming. I might change this naming later.

File: shared/src/main/java/com/couchbase/lite/ResultSet.java (hunk 4 above)
Comment (hideki) [inline shared/src/main/java/com/couchbase/lite/ResultSet.java]: My comment means `class ResultSet implements Iterable`. 
```java
ResultSet resultSet = ...;
for(QueryRow row: resultSet){
   ...
}
```

File: shared/src/main/java/com/couchbase/lite/Expression.java (hunk 3 above)
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Expression.java]: That is fine. Please note that I have seen lhs and rhs used before outside couchbase. Also apple also used this in its NSPredicate API (See https://developer.apple.com/reference/foundation/nscomparisonpredicate/1568141-predicatewithleftexpression?language=objc).

**Comment by hideki** [comment]
All spec related stuff, I created tickets in API repo.

File: shared/src/main/java/com/couchbase/lite/Expression.java (hunk 3 above)
Comment (hideki) [inline shared/src/main/java/com/couchbase/lite/Expression.java]: It seems lhs and rhs are commonly used. I will add comments like NSPredicate API.

**Pull request merged**